sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,290 @@
+  }
+}
+
+AZ_NODISCARD az_mqtt_options az_mqtt_options_default()
+{
+  return (az_mqtt_options){ .certificate_authority_trusted_roots = AZ_SPAN_EMPTY,
+                            .openssl_engine = NULL,
+                            .mosquitto_handle = NULL };
+}
+
+AZ_NODISCARD az_result az_mqtt_init(az_mqtt* mqtt, az_mqtt_options const* options)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt);
+  mqtt->_internal.options = options == NULL ? az_mqtt_options_default() : *options;
+  mqtt->_internal.mosquitto_handle = mqtt->_internal.options.mosquitto_handle;
+  mqtt->_internal.platform_mqtt.pipeline = NULL;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt_outbound_connect(az_mqtt* mqtt, az_mqtt_connect_data* connect_data)
+{
+  az_result ret = AZ_OK;
+  az_mqtt* me = (az_mqtt*)mqtt;
+
+  // IMPORTANT: application must call mosquitto_lib_init() before any Mosquitto clients are created.
+
+  if (me->_internal.mosquitto_handle == NULL)
+  {
+    me->_internal.mosquitto_handle = mosquitto_new(
+        (char*)az_span_ptr(connect_data->client_id),
+        false, // clean-session
+        me); // callback context - i.e. user data.
+  }
+
+  if (me->_internal.mosquitto_handle == NULL)
+  {
+    ret = AZ_ERROR_OUT_OF_MEMORY;
+    return ret;
+  }
+
+  // Configure callbacks. This should be done before connecting ideally.
+  mosquitto_log_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_log);
+  mosquitto_connect_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_connect);
+  mosquitto_disconnect_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_disconnect);
+  mosquitto_publish_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_publish);
+  mosquitto_subscribe_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_subscribe);
+  mosquitto_unsubscribe_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_unsubscribe);
+  mosquitto_message_callback_set(me->_internal.mosquitto_handle, _az_mosquitto_on_message);
+
+  // If the application has provided a CA root certificate, then use it. Otherwise, use the OS
+  // certs.
+  if (az_span_size(me->_internal.options.certificate_authority_trusted_roots) != 0)
+  {
+    _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_tls_set(
+        me->_internal.mosquitto_handle,
+        (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots),
+        NULL,
+        (const char*)az_span_ptr(connect_data->certificate.cert),
+        (const char*)az_span_ptr(connect_data->certificate.key),
+        NULL)));
+  }
+  else
+  {
+    _az_RETURN_IF_FAILED(_az_result_from_mosq(
+        mosquitto_int_option(me->_internal.mosquitto_handle, MOSQ_OPT_TLS_USE_OS_CERTS, 1)));
+    // Passing any string to mosquitto_tls_set() will use the OS certs when
+    // MOSQ_OPT_TLS_USE_OS_CERTS is set.
+    _az_RETURN_IF_FAILED(_az_result_from_mosq(
+        mosquitto_tls_set(me->_internal.mosquitto_handle, NULL, "L", NULL, NULL, NULL)));
```

Comment:
Nit: please #define "L" to mean REQUIRED_TLS_SET_CERT_PATH.
When you have some time, please file a bug with Mosquitto MQTT to describe this as a bug and ask if they could fix it so that we can pass in NULL (or, if necessary, some other #defined value).